src/helpers/keyUtils.ts

```diff
@@ -29,22 +44,219 @@ export function toChecksumAddress(hexAddress: string): string {
+
+  const finalUserKeyPair = secp256k1Curve.keyFromPrivate(scalar.toString("hex", 64));
+
+  return {
+    oAuthKeyScalar: oAuthKeyPair.getPrivate(),
+    oAuthPubX: oAuthPubKey.getX(),
+    oAuthPubY: oAuthPubKey.getY(),
+    SigningPubX: oAuthPubKey.getX(),
+    SigningPubY: oAuthPubKey.getY(),
+    metadataNonce: randomNonce,
+    encryptedSeed: "",
+    metadataSigningKey: oAuthKeyPair.getPrivate(),
+    finalUserPubKeyPoint: finalUserKeyPair.getPublic(),
+  };
+};
+
+export function generateAddressFromPrivKey(keyType: KeyType, privateKey: BN): string {
+  const ecCurve = getKeyCurve(keyType);
   const key = ecCurve.keyFromPrivate(privateKey.toString("hex", 64), "hex");
-  const publicKey = key.getPublic().encode("hex", false).slice(2);
-  const evmAddressLower = `0x${keccak256(Buffer.from(publicKey, "hex")).slice(64 - 38)}`;
-  return toChecksumAddress(evmAddressLower);
+  if (keyType === "secp256k1") {
+    const publicKey = key.getPublic().encode("hex", false).slice(2);
+    const evmAddressLower = `0x${keccak256(Buffer.from(publicKey, "hex")).slice(64 - 38)}`;
+    return toChecksumAddress(evmAddressLower);
+  } else if (keyType === "ed25519") {
+    const publicKey = encodeEd25519Point(key.getPublic());
+    const address = base58.encode(publicKey);
+    return address;
+  }
+  throw new Error(`Invalid keyType: ${keyType}`);
 }
 
-export function generateAddressFromPubKey(ecCurve: ec, publicKeyX: BN, publicKeyY: BN): string {
+export function generateAddressFromPubKey(keyType: KeyType, publicKeyX: BN, publicKeyY: BN): string {
+  const ecCurve = getKeyCurve(keyType);
   const key = ecCurve.keyFromPublic({ x: publicKeyX.toString("hex", 64), y: publicKeyY.toString("hex", 64) });
-  const publicKey = key.getPublic().encode("hex", false).slice(2);
-  const evmAddressLower = `0x${keccak256(Buffer.from(publicKey, "hex")).slice(64 - 38)}`;
-  return toChecksumAddress(evmAddressLower);
+  if (keyType === "secp256k1") {
+    const publicKey = key.getPublic().encode("hex", false).slice(2);
+    const evmAddressLower = `0x${keccak256(Buffer.from(publicKey, "hex")).slice(64 - 38)}`;
+    return toChecksumAddress(evmAddressLower);
+  } else if (keyType === "ed25519") {
+    const publicKey = encodeEd25519Point(key.getPublic());
+    const address = base58.encode(publicKey);
+    return address;
+  }
+  throw new Error(`Invalid keyType: ${keyType}`);
 }
 
-export function getPostboxKeyFrom1OutOf1(ecCurve: ec, privKey: string, nonce: string): string {
+export function getPostboxKeyFrom1OutOf1(ecCurve: EC, privKey: string, nonce: string): string {
   const privKeyBN = new BN(privKey, 16);
   const nonceBN = new BN(nonce, 16);
   return privKeyBN.sub(nonceBN).umod(ecCurve.curve.n).toString("hex");
 }
+
+export function derivePubKey(ecCurve: EC, sk: BN): curve.base.BasePoint {
+  const skHex = sk.toString(16, 64);
+  return ecCurve.keyFromPrivate(skHex).getPublic();
+}
+
+export const encryptionEC = new EC("secp256k1");
```

Comment:
pls don't define exportable variables. not good for treeshaking

Reply:
done